app/controllers/api/migrations.php

```diff
@@ -0,0 +1,824 @@
+    ->inject('dbForProject')
+    ->inject('dbForConsole')
+    ->inject('project')
+    ->inject('user')
+    ->inject('events')
+    ->inject('request')
+    ->action(function (array $resources, string $projectId, Response $response, Database $dbForProject, Database $dbForConsole, Document $project, Document $user, Event $eventsInstance, Request $request) {
+        $firebase = new OAuth2Firebase(
+            App::getEnv('_APP_MIGRATIONS_FIREBASE_CLIENT_ID', ''), 
+            App::getEnv('_APP_MIGRATIONS_FIREBASE_CLIENT_SECRET', ''), 
+            $request->getProtocol() . '://' . $request->getHostname() . "/v1/migrations/firebase/redirect"
+        );
+
+        $accessToken = $user->getAttribute('migrationsFirebaseAccessToken');
+        $refreshToken = $user->getAttribute('migrationsFirebaseRefreshToken');
+        $accessTokenExpiry = $user->getAttribute('migrationsFirebaseAccessTokenExpiry');
+
+        $isExpired = new \DateTime($accessTokenExpiry) < new \DateTime('now');
+        if ($isExpired) {
+            $firebase->refreshTokens($refreshToken);
+
+            $accessToken = $firebase->getAccessToken('');
+            $refreshToken = $firebase->getRefreshToken('');
+
+            $verificationId = $firebase->getUserID($accessToken);
+
+            if (empty($verificationId)) {
+                throw new Exception(Exception::GENERAL_RATE_LIMIT_EXCEEDED, "Another request is currently refreshing OAuth token. Please try again.");
+            }
+
+            $user = $user
+                ->setAttribute('migrationsFirebaseAccessToken', $accessToken)
+                ->setAttribute('migrationsFirebaseRefreshToken', $refreshToken)
+                ->setAttribute('migrationsFirebaseAccessTokenExpiry', DateTime::addSeconds(new \DateTime(), (int)$firebase->getAccessTokenExpiry('')));
+
+            $dbForConsole->updateDocument('users', $user->getId(), $user);
+        }
+        
+        $firebase->createServiceAccount($accessToken, $projectId);
+        
+        // $migration = $dbForProject->createDocument('migrations', new Document([
+        //     '$id' => ID::unique(),
+        //     'status' => 'pending',
+        //     'stage' => 'init',
+        //     'source' => Firebase::getName(),
+        //     'credentials' => [
+        //         'serviceAccount' => $serviceAccount,
+        //     ],
+        //     'resources' => $resources,
+        //     'statusCounters' => '{}',
+        //     'resourceData' => "{}",
+        //     'errors' => []
+        // ]));
+
+        // $eventsInstance->setParam('migrationId', $migration->getId());
+
+        // // Trigger Transfer
+        // $event = new Migration();
+        // $event
+        //     ->setMigration($migration)
+        //     ->setProject($project)
+        //     ->setUser($user)
+        //     ->trigger();
+
+        // $response
+        //     ->setStatusCode(Response::STATUS_CODE_ACCEPTED)
+        //     ->dynamic($migration, Response::MODEL_MIGRATION);
```

Comment:
Do we need this code or can it be removed?

Reply:
This code was commented out as I was still working on the code that generates a service account and didn't want to start a transfer everytime

Reply:
It will be uncommented once I'm done with that